### PR TITLE
Issue 43 password file creation command

### LINKED ISSUE #43: PasswordFileCreation tool to generate a file based user database for PasswordAuthHandler
PasswordFileCreation tool needs to be build based on this class: https://github.com/pravega/pravega/blob/master/controller/src/test/java/io/pravega/controller/auth/PasswordFileCreatorTool.java
to generate a file containing user data, in the same format as this file:
https://github.com/pravega/pravega/blob/master/config/passwd

This generated file can be used as user database for Pravega installation using basic PasswordAuthHandler. See [doc ](https://github.com/pravega/pravega-operator/blob/master/doc/auth.md) for details.

### PR DESCRIPTION
**Pull request by ldr7**
**Change log description**
PasswordFileCreation tool to generate a file based user database for PasswordAuthHandler

**Purpose of the change**
Fixes #43 

**What the code does**

Creates a command that takes as argument the name of the file to be generated and user details (username:password:acl). Encrypted password is generated from the user details. It then creates a file with filename given as argument and writes the encrypted password to the file. 

Registering the command in the ImmutableMap 'builder'.

Test file for the command

How to verify it
List the commands available in Pravega CLI. Use the new command and check its functionality.

### CODE REVIEW

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import io.pravega.tools.pravegacli.commands.Command;
+import io.pravega.controller.server.rpc.auth.StrongPasswordProcessor;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.List;
+
+/** The command takes filename and user:password:acl as argument. It creates a file and writes the encrypted password to the file.
+ *
```
Comment (RaulGracia) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: Please, format the comment correctly.

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ */
+public class PasswordFileCreatorCommand extends Command {
+    public PasswordFileCreatorCommand(CommandArgs args){super(args);}
```
Comment (RaulGracia) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: Please, place the constructor (without collapsing the code), after the field declarations.

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    }
+
+    public String getTargetFilename(List<String> userInput) {
+        return userInput.get(0);
+    }
+
+    public String getUserDetails(List<String> userInput){
+        String userDetails = userInput.get(1);
+        if((userDetails.split(":")).length == 3){
```
Comment (RaulGracia) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: Missing space afer `if` and before `(`, as well as after `)` and before `{`.

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            return userDetails;
+        }
```
Comment (RaulGracia) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: Please, write `else` blocks as: `} else {`

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    public void createPassword(String userDetails) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        String[] lists = parseUserDetails(userDetails);
+        toWrite = generatePassword(lists);
+    }
+
+    private String[] parseUserDetails(String userDetails){
+        return userDetails.split(":");
+    }
+
+    private String generatePassword(String[] lists) throws NoSuchAlgorithmException, InvalidKeySpecException{
```
Comment (RaulGracia) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: Missing space after `InvalidKeySpecException` and before `{`.

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    private void writeToFile(String targetFileName, String toWrite) throws IOException {
+        try (FileWriter writer = new FileWriter(targetFileName))
+        {
+            writer.write(toWrite + "\n");
+            writer.flush();
+        }
+    }
+
+    public static CommandDescriptor descriptor() {
+        final String COMPONENT = "admin";
```
Comment (RaulGracia) [inline pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: This is a local variable, so use camel case to define it. By the way, what is it actually doing? why it is hardcoded to `admin`?

File: pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        String s2 = "testing:testing:testing";
+        arguments.add(s1);
+        arguments.add(s2);
+        CommandArgs args = new CommandArgs(arguments, STATE);
+        PasswordFileCreatorCommand testObj = new PasswordFileCreatorCommand(args);
+        Assert.assertTrue(testObj.getTargetFilename(arguments).equals("TestFile"));
+    }
+
+    @Test
+    public void testgetUserDetails() throws Exception{
```
Comment (RaulGracia) [inline pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java]: Missing space after `Exception` and before `{ `.

File: pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            testObj.createPassword("testing:testing:testing");
+            Assert.assertTrue(testObj.getToWrite().split(":")[0].equals("testing"));
+            Assert.assertTrue(testObj.getToWrite().split(":")[2].equals("testing;"));
+        } catch(Exception e){
+            System.err.println(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testgetTargetFilename() throws Exception{
```
Comment (RaulGracia) [inline pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java]: Missing space after `Exception` and before `{`.

File: pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import org.junit.Rule;
+import java.io.File;
+import java.nio.file.*;
+
+public class PasswordFileCreatorCommandTest {
+
+    @Test
+    public void testcreatePassword() throws Exception {
+        try {
+            AdminCommandState STATE = new AdminCommandState();
```
Comment (RaulGracia) [inline pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java]: If this is a local variable, use camel case to define it, not capitals.

**Review by RaulGracia — CHANGES_REQUESTED** [review]

File: pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java (hunk 6 above)
Comment (ldr7) [reply pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: This has the same role as` bk` for all bookkeeper commands. It is part of the `CommandDescriptor` for this command.

Comment (RaulGracia) [reply pravega-cli/src/main/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommand.java]: In all the other commands, this is a `protected` field in an `abstract` class. You are not following this pattern in this PR, but at least define this as an attribute in the class, not as a local variable in capitals (it is confusing).

**Comment by RaulGracia** [comment]
@jdmaguire @sarlaccpit can you please verify if the command does what is expected from your side?

File: pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java (hunk 9 above)
Comment (ldr7) [reply pravega-cli/src/test/java/io/pravega/tools/pravegacli/commands/admin/PasswordFileCreatorCommandTest.java]: Done

**Comment by jdmaguire** [comment]
Nothing unusually seems to occur.

**Review by RaulGracia — APPROVED** [review]
Tested and it works, and seems to meet the expectations from platform.

**Pull request merged**